docs/tutorial.md

```diff
@@ -323,3 +323,449 @@ as well as the C++ standard library and increases the dependency surface.
+detect the device connection and disconnection by listening to udev
+events, but a supported device may also be not directly accessible from
+the OS --for example, a composite device that contains an updatable chip
+that’s connected through I2C to a USB hub that acts as an interface. In
+that case, the device discovery and enumeration must be programmed by
+the developer, but the same device identification and management
+mechanisms apply in all cases. See the “Creating a new device type” and
+“Device identification” below for more details.
+
+#### If the firmware type is not supported
+
+Same as with the new device type, there could be an existing firmware
+type that can be used as a base type for our new type, so first of all
+we should look for firmware types that are similar to the one we’re
+using. Then, choosing where to define the new type depends on whether
+there’s already a plugin that implements most of the functionalities we
+need or not.
+
+### Example: extending a firmware type
+
+Our firmware files are Intel HEX files that have optional
+vendor-specific sections at fixed addresses, this is not supported by
+any firmware type in fwupd out of the box but the [FuIhexFirmare](https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/fu-ihex-firmware.c) class
+parses and models a standard Intel HEX file, so we can create a subclass
+of it for our firmware type and override the parse method so that it
+calls the method from the parent class, which would parse the file, and
+then we can get the data with `fu_firmware_get_bytes()` and do the rest of
+the custom parsing. Example:
+https://github.com/fwupd/fwupd/blob/main/plugins/analogix/fu-analogix-firmware.c
+
+### Example: extending a device type
+
+Communication with our new device is carried out by doing
+read/write/ioctl operations on a device file, but using a custom
+protocol that is not supported in fwupd.
+
+For this type of device we can create a new type derived from
+`FuUdevDevice`, which takes care of discovering this type of devices,
+possibly using a vendor-specific protocol, as well as of opening,
+reading and writing device files, so we’d only have to implement the
+protocol on top of those primitives. (Example:
+`fu_logitech_hidpp_runtime_bolt_poll_peripherals()` in
+https://github.com/fwupd/fwupd/blob/main/plugins/logitech-hidpp/fu-logitech-hidpp-runtime-bolt.c)
+The process would be similar if our device was handled by a different
+backend (USB or BlueZ).
+
+### Creating a new plugin
+
+The bare minimum a plugin should have is a `fu_plugin_init` function that
+defines the plugin characteristics such as the device type and firmware
+type handled by it, the build hash and any plugin-specific quirk keys
+that can be used for the plugin.
+
+    void
+    fu_plugin_init(FuPlugin *plugin)
+    {
+        FuContext *ctx = fu_plugin_get_context(plugin);
+
+        fu_plugin_set_build_hash(plugin, FU_BUILD_HASH);
```

Comment:
i think this is different now in main, you set the build hash in the vfunc setup.